misk-feature-testing/src/main/kotlin/misk/feature/testing/FakeFeatureFlags.kt

```diff
@@ -77,76 +77,173 @@ class FakeFeatureFlags @Inject constructor(private val moshi : Provider<Moshi>)
   override fun <T : Enum<T>> getEnum(feature: Feature, clazz: Class<T>): T = getEnum(feature, KEY, clazz)
   override fun <T> getJson(feature: Feature, clazz: Class<T>): T = getJson(feature, KEY, clazz)
 
-  private fun get(feature: Feature, key: String): Any? {
+  private fun <V> getOrDefault(
+    feature: Feature,
+    key: String,
+    defaultValue: V,
+    attributes: Attributes = defaultAttributes
+  ): V {
     FeatureFlagValidation.checkValidKey(feature, key)
-    return overrides.getOrElse(MapKey(feature, key)) {
-      overrides[MapKey(feature)]
-    }
+    @Suppress("unchecked_cast")
+    return (get(feature, key, attributes) ?: defaultValue) as V
   }
 
-  private fun <V> getOrDefault(feature: Feature, key: String, defaultValue: V): V {
+  private fun get(
+    feature: Feature,
+    key: String,
+    attributes: Attributes = defaultAttributes
+  ): Any? {
     FeatureFlagValidation.checkValidKey(feature, key)
-    @Suppress("unchecked_cast")
-    return overrides.getOrElse(MapKey(feature, key)) {
-      overrides.getOrDefault(MapKey(feature), defaultValue)
-    } as V
+    // Override value lookup is structured in 3 levels: feature, key, and attributes level. The
+    // logic to get the override value is as follows:
+    //
+    // 1. Return feature level override if there is no override at the key level.
+    // 2. if there is an override at the key level, then:
+    //   2.1 Return the override value associated with the attribute override that contains
+    //       all the attributes of the given attributes.
+    //   2.2 If there is no match, return the override value defined at the key level.
+    val overrideMapValues = overrides[MapKey(feature, key)]
+        ?: return overrides[MapKey(feature)]?.let { it.first().value }
+
+    val overrideMapValuesCopy = PriorityQueue(overrideMapValues)
+
+    var currentMapValue = overrideMapValuesCopy.poll()
+    while(overrideMapValuesCopy.isNotEmpty()) {
+      if (attributes.text.entries.containsAll(currentMapValue.attributes.text.entries)) { break }
+      currentMapValue = overrideMapValuesCopy.poll()
+    }
+    return currentMapValue.value
```

Comment:
is there a test that exercises this line? there are overrides for attributes, and none match?

Reply:
this is probably the line I understand the least . . . the attributes in this result won’t match, right?

Reply:
> is there a test that exercises this line? there are overrides for attributes, and none match?

Yes, I defined the following test:

```
    // Override key level
    subject.overrideKeyJson(FEATURE, "joker", JsonFeature("joker"))

    //override key and attributes level
    val badJokerAttributes = Attributes(mapOf("type" to "bad"))
    subject.overrideKeyJson(FEATURE, "joker", JsonFeature("bad-joker"), badJokerAttributes)
```

On exact attributes match `badJokerAttributes`, return `JsonFeature("bad-joker")`:
```
    assertThat(subject.getJson<JsonFeature>(FEATURE, "joker", badJokerAttributes))
        .isEqualTo(JsonFeature("bad-joker"))
```

On subtype attribute match, notice that `sleepyBadJokerAttributes` has one more attribute than `badJokerAttributes`, return `JsonFeature("bad-joker")`:

```
    val sleepyBadJokerAttributes = Attributes(mapOf("type" to "bad", "state" to "sleepy"))
    assertThat(subject.getJson<JsonFeature>(FEATURE, "joker", sleepyBadJokerAttributes))
        .isEqualTo(JsonFeature("bad-joker"))
```

On none attribute match, return `JsonFeature("joker")` which is the key level override:

```
    val goodJokerAttributes = Attributes(mapOf("type" to "good"))
    assertThat(subject.getJson<JsonFeature>(FEATURE, "joker", goodJokerAttributes))
        .isEqualTo(JsonFeature("joker"))
```

This test is defined here: https://github.com/cashapp/misk/pull/1565/files#diff-e2b08dff1b3751d173f3a21aa99d695aR116-R132

Reply:
> this is probably the line I understand the least . . . the attributes in this result won’t match, right?

So the main idea is that you loop through all the attributes we have in the overrides. If a match is found, the while loop breaks and the match is returned. There is the case that the while loop won't ever break but it will end when all the overrides have been consumed. If you ever reached that case, then you are in the situation where none of the attributes were a match, so the key level override will be returned. The logic is commented here in the code: https://github.com/cashapp/misk/pull/1565/files#diff-89dcbd7609483933353dac8a90bd2ca4R97-R104 (see the line with `2.2` in the front).